hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsErrorVisibilityHandler.kt

```diff
@@ -0,0 +1,66 @@
+package uk.gov.justice.hmpps.sqs
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.awspring.cloud.sqs.listener.QueueMessageVisibility
+import io.awspring.cloud.sqs.listener.SqsHeaders
+import org.slf4j.LoggerFactory
+import org.springframework.messaging.Message
+import software.amazon.awssdk.services.sqs.model.MessageAttributeValue
+
+class HmppsErrorVisibilityHandler(
+  private val objectMapper: ObjectMapper,
+  private val hmppsSqsdProperties: HmppsSqsProperties,
+) {
+  fun setErrorVisibilityTimeout(message: Message<Any>, queueId: String) {
+    val eventType = getEventType(message)
+    val receiveCount = (message.headers["Sqs_Msa_ApproximateReceiveCount"] as String).toInt()
+    val sqsVisibility = message.headers["Sqs_VisibilityTimeout"] as QueueMessageVisibility
+
+    val timeouts = hmppsSqsdProperties.queues[queueId]
+      ?.let { queue ->
+        queue.eventErrorVisibilityTimeout?.get(eventType)
+          ?: queue.errorVisibilityTimeout
+      }
+      ?: hmppsSqsdProperties.defaultErrorVisibilityTimeout
+
+    val nextTimeoutSeconds = when {
+      receiveCount <= timeouts.size -> timeouts[receiveCount-1]
+      timeouts.isNotEmpty() -> timeouts.last()
+      else -> 0
+    }
+    log.info("Setting error visibility timeout for event type {} on queue {} with receive count {} to {} seconds", eventType, queueId, receiveCount, nextTimeoutSeconds)
+    sqsVisibility.changeTo(nextTimeoutSeconds)
+  }
```

Comment:
The original plan was to store the number of retries and timeout values in message attributes. It turns out this isn't possible in the error handler.

Instead we know the retry count from the AWS header `Sqs_Msa_ApproximateReceiveCount`, so we can calculate the next timeout from that and the configuration properties.